### PR TITLE
TopologySpreadConstraint improvements

### PR DESCRIPTION
**Pull request by aermakov-zalando**
The upstream TopologySpreadConstraint is pretty much unusable in even the medium-sized clusters we run, and it definitely won't work during the load tests or other major events. There are some fixes upstream, which seem to mitigate the performance issues, but make CA overshoot the scaling target by a factor of 20x sometimes. I don't really want to spend more time digging in this amazing codebase and figuring out how to fix this and not break everything else, so let's do something simpler instead. We'll recognise the automatically-injected topology spread constraints we add in the admission controller and emulate them using much simpler logic instead, by rewriting them internally into simple `nodeSelector` conditions. This seems to work insanely fast, and because the scheduler isn't affected that horribly, the end result might still be usable.

In addition to that, we'll also add a safety catch in place. Upstream topology spread constraint implementation is a binary thing, which is either in place or not (`whenUnsatisfiable` is pretty much useless in an autoscaling cluster like a lot of other scheduling features), but this might be dangerous during zone outages or capacity shortages. Since we're emulating the whole thing anyway, this PR adds support for a special annotation (`zalando.org/topology-spread-timeout`) that enables the predicate for the first X minutes after a pod has been created, and then completely ignores it after the timeout expires. This is paired with a similar fix to the scheduler, and should allow the pod to still run _somewhere_ in case of scaling issues, even if the resulting distribution would be less than ideal.

The PR is best read commit-by-commit, I can also split it if needed. Changes:
 * revert the `TopologySpreadConstraintSplitFactor` feature, it's not really needed anymore
 * introduce the TSC emulation for the constraint that we inject in the admission controller (optional, can be disabled)
 * simulation tests: add basic support for scale-down (evicted pods are deleted immediately, anything different will be much harder to implement properly, but even this basic implementation is enough to test simple scenarios)
 * allow the emulated TSC constraint to time out if configured via an annotation

### CODE REVIEW

File: cluster-autoscaler/core/tsc_emulation.go [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+	}
+
+	var result []*corev1.Pod
+	for _, pod := range original {
+		if hasEmulatedTopologySpreadConstraints(pod, l.constraintLabel) {
+			pod = pod.DeepCopy()
+			pod.Spec.TopologySpreadConstraints = nil
+
+			// Pin already scheduled pods to their current zones, to avoid potential issues with downscaling
+			if nodeZone, ok := l.nodeZoneMapping[pod.Spec.NodeName]; ok && nodeZone != "" {
```
Comment (mikkeloscar) [inline cluster-autoscaler/core/tsc_emulation.go:94]: Do we consider `Succeed/Failed` pods here? I believe those also have the `nodeName` set which could result in wrong distribution if we don't filter them out somewhere else?

Comment (aermakov-zalando) [reply cluster-autoscaler/core/tsc_emulation.go:94]: This wraps a [`scheduledPodLister`](https://github.com/zalando-incubator/autoscaler/blob/0e73b3e53035bfb05d0af1290d099895fe121dc7/cluster-autoscaler/utils/kubernetes/listers.go#L207), which already filters those out.

Comment (aermakov-zalando) [reply cluster-autoscaler/core/tsc_emulation.go:94]: (no idea what's going on with the selector there, don't ask!)

**Comment by mikkeloscar** [comment]
:+1:

**Comment by linki** [comment]
👍

**Pull request merged**